# Exercise 2.A Fundamentals of pandas

In [48]:
import numpy as np
import pandas as pd

## Series

In [49]:
# Var aa, ab, ac, ad
aa = [1,2,3,4]
ab = (5,6,7,8)
ac = {9,10,11,12}
ad = {"first": 13, "second": 14, "third": 15,"fourth": 16}

In [50]:
# Var ser1-4
ser1 = aa
ser2 = ab
ser3 = ac
ser4 = ad

In [51]:
ser5 = (17, 18, 19, 20) # No Errors ¯\_(ツ)_/¯

## DataFrames From Series

In [52]:
df1 = pd.DataFrame(ser1)
df1

,0
0,1
1,2
2,3
3,4


In [53]:
# df1.insert(1, "column2", ser2)
df1.insert(0, "column2", ser2)
# df1.insert(2, "column3", ser3) #Cannot add dictionary to dataframe using insert
df1

,column2,0
0,5,1
1,6,2
2,7,3
3,8,4


In [60]:
serv4 = pd.Series(ad)
df1["Column4"] = ser4
### RETURNS NaN#####

In [61]:
df1 

,column2,0,Column4
0,5,1,NaN
1,6,2,NaN
2,7,3,NaN
3,8,4,NaN


In [66]:
df1.loc[4] = [5, 9, 12,0]
##VALUE ERROR???!?!?!??!?###

ValueError: Must have equal len keys and value when setting with an iterable

## Reading Data into a DataFrame (Hopefully I have more luck here)

In [81]:
df2 = pd.read_csv('sakila_example.csv')
df2

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE
...,...,...,...,...,...
257,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,GARY PENN
258,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,CHRISTOPHER WEST
259,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,DUSTIN TAUTOU
260,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,ANGELA HUDSON


In [82]:
df2.head(5)


,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE


In [83]:
df2.tail(10)

,title,description,category,language,actor_name
252,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,MILLA KEITEL
253,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,GROUCHO DUNST
254,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,BURT TEMPLE
255,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,UMA WOOD
256,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,RIP WINSLET
257,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,GARY PENN
258,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,CHRISTOPHER WEST
259,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,DUSTIN TAUTOU
260,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,ANGELA HUDSON
261,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,JAMES PITT


In [85]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 262 entries, 0 to 261
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   title        262 non-null    object
 1   description  262 non-null    object
 2   category     262 non-null    object
 3   language     262 non-null    object
 4   actor_name   262 non-null    object
dtypes: object(5)
memory usage: 10.4+ KB


From the info results I can see that there are five columns within the csv.
title, description, category, language, and actor name. The data type for each is object.


In [87]:
df3 = pd.read_csv('northwind_example.csv')
df3

,OrderID,ProductName,UnitPrice,Quantity
0,10248,Queso Cabrales,14.0,12
1,10248,Singaporean Hokkien Fried Mee,9.8,10
2,10248,Mozzarella di Giovanni,34.8,5
3,10249,Manjimup Dried Apples,42.4,40
4,10249,Tofu,18.6,9
...,...,...,...,...
495,10435,Mozzarella di Giovanni,27.8,10
496,10436,Gnocchi di nonna Alice,30.4,40
497,10436,Wimmers gute Semmelkndel,26.6,30
498,10436,Rhnbru Klosterbier,6.2,24


In [88]:
df3.describe()

,OrderID,UnitPrice,Quantity
count,500.000000,500.000000,500.000000
mean,10341.364000,23.371400,24.330000
std,54.442663,27.334436,18.069702
min,10248.000000,2.000000,1.000000
25%,10294.000000,10.400000,10.000000
50%,10341.000000,15.800000,20.000000
75%,10389.000000,27.800000,32.000000
max,10436.000000,210.800000,120.000000


From the describe results I can gather that there are 500 orders. 
The other values are as follows:
Standard Deviation: 54.44
Minimum Price: 2.00
Maximum Price: 210.

The summarizations are different because the nature of the data is different.

## Coding A DataFrame

In [90]:
test_data = {
    'ID': [1, 2, 3, 4, 5, 6, 7, 8],
    'Name': ['Alicia', 'Bob', 'Charlie', 'Dominic', 'Eve', 'Frank', 'Grace', 'Heidi'],
    'Age': [25, np.nan, 22, 19, 31, 35, np.nan, 18],
    'City': ['NY', 'LA', 'Chicago', 'Houston', 'Phoenix', 'Boston', 'Austin', 'San Diego'],
    'Score': [85, 92, -1, 65, 78, np.nan, 55, 90]
}

In [91]:
df4 = test_data
df4

{'ID': [1, 2, 3, 4, 5, 6, 7, 8],
 'Name': ['Alicia',
  'Bob',
  'Charlie',
  'Dominic',
  'Eve',
  'Frank',
  'Grace',
  'Heidi'],
 'Age': [25, nan, 22, 19, 31, 35, nan, 18],
 'City': ['NY',
  'LA',
  'Chicago',
  'Houston',
  'Phoenix',
  'Boston',
  'Austin',
  'San Diego'],
 'Score': [85, 92, -1, 65, 78, nan, 55, 90]}

In [93]:
df4 = pd.DataFrame(test_data)
df4.to_csv('test_data.csv')